### PR TITLE
Add _parse_component for supply() and facilities()

### PR DESCRIPTION
**Pull request by prior-l**
A refactor to simply the logic in supply(), facilities(), moon_facilities(), research(), defense(), ships() to a single place.

If you accept this I'll convert research(), defense(), ships() in a follow up.

I tested this works
Screenshot from 2021-02-17 22-39-39

### CODE REVIEW

File: ogame/__init__.py [hunk 1]
```diff
@@ -273,77 +273,106 @@ def collect_status(self):
...
-                self.level = levels[i]
-                self.data = OGame.collect_status(status[i])
+            def __init__(self, level, status):
+                self.level = level
+                self.data = OGame.collect_status(status)
                 self.is_possible = self.data[0]
                 self.in_construction = self.data[1]
 
         class Supplies(object):
-            metal_mine = Supply(0)
```
Comment (prior-l) [inline ogame/__init__.py:290]: Currently it's hard to fleetsave because it's difficult to convert the `class Ships` from `ships()` to a list of ships and counts.

This change adds a `count` field to each return class. Later it seems like these would benefit from being converted to NamedTuples at which point the 6 methods (ships, defense, research, buildings, facilities, moon_facilities) become one lines

```
def Supply():
     return self._parse_component('supplies', const.building_supply_names)

def Defense():
     return self._parse_component('defenses', const.building_defense_names)
```

**Comment by PiecePaperCode** [comment]
i didnt know you could dynamicly add variables to a class by simply calling self.__dict__ its super cool 

as per screenshot the class keeps its Supply.metal_mine calling afterwards what would make it possible to ditch the inheritance Model.

Yes its hard to parse Ships to ships because they dont share the same exact class. :(. As a Miner myself its my fingerprint. but its tricky because you need to know ships who can fly and those who dont like satellite and transporters.

can you give parce_component a bedder name. it souns so generic. Maby Parse Strucktures.
count could be levels levelsList

**Comment by prior-l** [comment]
Done.

**Comment by PiecePaperCode** [comment]
_Just want to say i dont really like ToDo Comments because the normaly communicate to me a "dont do"._
Now a Problem arosed when trying to work with your code. Because everything is dynamic now Pycharm fails to offer autocomplete. Is there a way to solve this. Most people using this lib are python beginners and will get frustrating by triying to write programms and the IDE fails to tell them if they made an spelling error.

```python
print(empire.supply(empire.planet_ids()[0]).deuterium_storage.in_constructions)
```
this code start up fine until it gets to it.
